rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+```
+
+The database client (such as psql) will talk to Teleport web proxy port (`3080`
+by default) which will use multiplexing to detect the database protocol and
+dispatch to an appropriate service.
+
+## Authentication
+
+In this model, there are 3 points where authentication needs to happen.
+
+### Database client <---> proxy
+
+**Mutual TLS.** Database clients, such as psql, will use short-lived x509
+certificates issued by Teleport in order to authenticate with the proxy.
+
+The certificate metadata includes (via extensions) routing information about
+the target database which users log into using `tsh db login` command (see
+below for UX).
+
+### Proxy <---> database service
+
+**Mutual TLS.** The connection that is passed from proxy to a database service
+is upgraded to TLS as well in order to be able to pass identity information
+over the reverse tunnel.
+
+In order to do that, proxy will generate an ephemeral certificate signed
+by the user authority and re-encode all identity and routing information
+extracted from the certificate supplied by the client into it.
+
+### Database service <---> database instance
+
+**Mutual TLS (onprem).**  When connecting to an onprem database instance, the
+database service will use a client certificate for authentication, which it
+will generate on the fly according to the database requirements (for example,
+PostgreSQL requires the database user name to be encoded in the certificate
+as common name).
+
+This also means that the database needs to be configured with Teleport's
+certificate authority and (in general case) use server certificate issued by
+Teleport which can be done using `tctl auth sign` command (see below for UX).
+
+**Password auth (RDS/Aurora).** Amazon RDS/Aurora don't support client cert
+authentication. Instead, they support IAM authentication so Teleport database
+service will generate a short-lived authentication token using RDS API that
+will be used as a password:
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html
+
+In order to be able to verify RDS/Aurora server certificate, users will need
+to download RDS root certificate and configure the database service to trust
+it:
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html
+
+## Configuration
+
+### Teleleport database service
+
+The following new configuration section is added to the Teleport file config:
+
+```yaml
+# New global key housing the database service configuration.
+db_service:
+  # Enable or disable the database service.
+  enabled: "yes"
+  # List of the database this service is proxying.
+  databases:
+    # Database instance name, used to refer to an instance in CLI like tsh.
+  - name: "postgres-prod"
+    # Optional free-form verbose description of a database instance.
+    description: "Production instance of PostgreSQL 13.0"
+    # Database procotol, only "postgres" is supported initially.
+    protocol: "postgres"
```

Comment:
@klizhentas @benarent Since we're calling the endpoint "uri" now, I've just got an idea that we maybe can also get rid of this separate "protocol" configuration field and instead just require a proper scheme for the URI e.g.

```
uri: "postgres://psql.internal.example.com:5432"
uri: "mysql://db.dev.gravitational.io:3306"
```

Would probably be more consistent as well since procotol is a part of uri normally, what do you think?

Reply:
@r0mant I think it should be consistent with what developers and devops are used to, how do connectors look like in DB SQL examples?

Reply:
For traditional relational databases like Postgres or MySQL this is actually exactly how their connection strings look like as it follows RFC3986 (URI spec). Looking at some other examples more modern databases and their drivers also use the same URI syntax like `mongodb://`, `redis://`, `cassandra://`, etc. so the syntax should definitely be familiar to devops folks and engineers. The only example off the top of my head where I haven't seen this scheme used before would be etcd, but I think if its URI looks like `etcd://1.2.3.4:4001` then it'd also be pretty clear and we wouldn't lose any information since we're only going to support https endpoint anyway.

Reply:
After some more thinking about this, decided to keep it as-is and have an explicit protocol field instead of specifying it as a uri scheme. The reason is that most dbs also support connections over unix domain sockets and normally it's specified as a scheme in the uri (i.e. `unix://` vs default `tcp://`). Don't know if that's a use case we'll want to support but I could see it potentially being useful (e.g. when running teleport as a side-car on the same node where a database is running).